infrastructure/lib/constructs/api-function.ts

```diff
@@ -45,13 +41,22 @@ export abstract class ApiFunction extends cdk.Construct {
    */
   public readonly route: apigw.Method;
 
+  /**
+   * The CfnFunction logicalId, used for templating the api yaml.
+   */
+  public readonly logicalId: lambda.CfnFunction;
+
   protected constructor(scope: cdk.Construct, id: string, props: ApiFunctionProps) {
     super(scope, id);
     this.method = props.method;
     this.fn = new lambdaNodeJs.NodejsFunction(this, "Function", {
       entry: props.handlerPath,
       ...props.functionPropsOverride,
     });
-    this.route = props.resource.addMethod(props.method, new apigw.LambdaIntegration(this.fn));
+    this.fn.addPermission("AllowApiGatwewayInvoke", {
+      principal: new iam.ServicePrincipal("apigateway.amazonaws.com"),
+    });
+    this.logicalId = this.fn.node.defaultChild as lambda.CfnFunction;
+    this.logicalId.overrideLogicalId(id);
```

Comment:
Using the ID here is clever and I see what you're doing but I don't know if it's actually super safe to do. I'd prefer to add it as an element of the `ApiFunctionProps` and to use that instead. First, it makes it far more explicit what it's being used for so there's less chance of someone forgetting that it has to match the spec file. Second, this opens up the possibility to a naming collision since the name of this attribute is now specifically set to `id` but the construct itself has the same name. Luckily our construct itself isn't raw CloudFormation. But it still makes me a little nervous.

If you've seen examples from AWS or other code bases where they will override the logical ID using the ID of the parent construct, then I will accept this as idiomatic; however, barring that, this is better as a required field in the prop. Perhaps named `functionLogicalId` or something like that.

Reply:
I am confused on what to change. The `id` is a string passed in:
```
const createPortfolioStep = new ApiDynamoDBFunction(this, "CreatePortfolioStepFunction", {
```

```
protected constructor(scope: cdk.Construct, id: string, props: ApiFunctionProps) {
```
^The id for this implementation would be `CreatePortfolioStepFunction`